File: Grind 75 in Java/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        dfs(new ArrayList<Integer>(), res, nums, new boolean[nums.length]);
        return res;
    }

    private void dfs(List<Integer> tmp, List<List<Integer>> res, int[] nums, boolean[] taken) {
        if (tmp.size() == nums.length) {
            res.add(List.copyOf(tmp));
            return;
        }

        for (int j = 0; j < nums.length; j++) {
            int n = nums[j];

            if (taken[j]) continue;

            tmp.addLast(n);
            taken[j] = true;

            dfs(tmp, res, nums, taken);

            taken[j] = false;
            tmp.removeLast();
        }
    }
}

File: leetcode in kotlin/0046-permutations.kt
class Solution {
    fun permute(nums: IntArray): List<List<Int>> {
        val res = ArrayList<ArrayList<Int>>()

        fun perm(i: Int) {
            if (i == nums.size) {
                res.add(nums.toCollection(ArrayList()))
                return
            }

            for (j in i until nums.size) {
                nums.swap(i, j)
                perm(i + 1)
                nums.swap(i, j)
            }
        }

        perm(0)
        return res
    }

    fun IntArray.swap(i: Int, j: Int) {
        this[i] = this[j].also { this[j] = this[i] }
    }
}
